activemq-kahadb-store/src/main/java/org/apache/activemq/store/kahadb/KahaDBStore.java

```diff
@@ -733,35 +734,56 @@ public void execute(Transaction tx) throws Exception {
         }
 
         @Override
-        public void recoverNextMessages(final int offset, final int maxReturned, final MessageRecoveryListener listener) throws Exception {
+        public void recoverMessages(final MessageRecoveryContext messageRecoveryContext) throws Exception {
             indexLock.writeLock().lock();
             try {
                 pageFile.tx().execute(new Transaction.Closure<Exception>() {
                     @Override
                     public void execute(Transaction tx) throws Exception {
                         StoredDestination sd = getStoredDestination(dest, tx);
+
+                        Long startSequenceOffset = null;
+                        Long endSequenceOffset = null;
+
+                        if(messageRecoveryContext.getStartMessageId() != null && !messageRecoveryContext.getStartMessageId().isBlank()) {
+                            startSequenceOffset = sd.messageIdIndex.get(tx, messageRecoveryContext.getStartMessageId());
+                        }
+
+                        if(startSequenceOffset == null) {
+                            startSequenceOffset = messageRecoveryContext.getOffset();
+                        }
+
+                        if(messageRecoveryContext.getEndMessageId() != null && !messageRecoveryContext.getEndMessageId().isBlank()) {
+                            endSequenceOffset = sd.messageIdIndex.get(tx, messageRecoveryContext.getEndMessageId());
+                        }
+
+                        if(endSequenceOffset == null) {
+                            endSequenceOffset = startSequenceOffset + Long.valueOf(messageRecoveryContext.getMaxMessageCountReturned());
+                        }
+
+                        if(endSequenceOffset < startSequenceOffset) {
+                            // Fast fail on invalid scenario
+                            return;
+                        }
+
+                        messageRecoveryContext.setEndSequenceId(endSequenceOffset);
                         Entry<Long, MessageKeys> entry = null;
-                        int position = 0;
-                        int counter = recoverRolledBackAcks(destination.getPhysicalName(), sd, tx, maxReturned, listener);
-                        Set ackedAndPrepared = ackedAndPreparedMap.get(destination.getPhysicalName());
-                        for (Iterator<Entry<Long, MessageKeys>> iterator = sd.orderIndex.iterator(tx); iterator.hasNext(); ) {
+                        recoverRolledBackAcks(destination.getPhysicalName(), sd, tx, messageRecoveryContext.getMaxMessageCountReturned(), messageRecoveryContext);
+                        Set<String> ackedAndPrepared = ackedAndPreparedMap.get(destination.getPhysicalName());
+                        Iterator<Entry<Long, MessageKeys>> iterator = (messageRecoveryContext.isUseDedicatedCursor() ? sd.orderIndex.iterator(tx, new MessageOrderCursor(startSequenceOffset)) : sd.orderIndex.iterator(tx));
```

Comment:
I'd add a new line here somewhere make it more readable as it's a long line but just a style thing so not a big deal

Reply:
done